mito-ai/src/utils/agentActions.tsx

```diff
@@ -0,0 +1,63 @@
+import { JupyterFrontEnd } from "@jupyterlab/application"
+import { CodeCell } from "@jupyterlab/cells"
+import { INotebookTracker } from "@jupyterlab/notebook"
+import { getFullErrorMessageFromTraceback } from "../Extensions/ErrorMimeRenderer/errorUtils"
+import { sleep } from "./sleep"
+import { didCellExecutionError } from "./notebook"
+import { ChatHistoryManager, PromptType } from "../Extensions/AiChat/ChatHistoryManager"
+
+export const acceptAndRunCode = async (
+    app: JupyterFrontEnd,
+    previewAICode: () => void,
+    acceptAICode: () => void,
+) => {
+    previewAICode()
+    acceptAICode()
+    await app.commands.execute("notebook:run-cell");
+}
+
+export const retryIfExecutionError = async (
+    notebookTracker: INotebookTracker, 
+    app: JupyterFrontEnd,
+    getDuplicateChatHistoryManager: () => ChatHistoryManager,
+    addAIMessageFromResponseAndUpdateState: (messageContent: string, promptType: PromptType, chatHistoryManager: ChatHistoryManager, mitoAIConnectionError?: boolean, mitoAIConnectionErrorType?: string | null) => void,
+    sendDebugErrorMessage: (errorMessage: string, agent?: boolean) => Promise<void>,
+    previewAICode: () => void,
+    acceptAICode: () => void,
+): Promise<boolean> => {
+    console.log('checking for error')
+    const cell = notebookTracker.currentWidget?.content?.activeCell as CodeCell;
+    const MAX_RETRIES = 3;
```

Comment:
Add comment. If we are going to update this then we need to update the message below